### ISSUE TITLE
Strange behaviour when mixing split_when with join_when extension

### ISSUE THREAD

**Issue by cjdb** [ericniebler/range-v3#1041]
## Problem

Given the use-case from my lexer (Snippet A), the snippet from `join_when_view<R, Pred>::iterator::next` (Snippet B) yields a logic error. Interestingly, when `lexeme == R"("\thell\\\o")"sv`, this logic error doesn't manifest.

Please see 1ef6154bb9d7e8a8c62385be75bff94278672760 for the `join_when_view` implementation. It's based off `v1.0-beta`.

## Snippet A
```cpp
{
   using namespace std::string_view_literals;
   constexpr auto legal_escapes = R"(bfnrt\'")"sv;
   auto is_escape = [legal_escapes](auto first, auto last) constexpr noexcept {
      auto result = false;
      if (auto next = ranges::next(first); next != last) {
         result = *first == '\\' and legal_escapes.find(*next) != std::string_view::npos;
      }

      return std::make_pair(result, first);
   };

   auto lexeme = R"(\thell\\\o)"sv;
   auto escapes = lexeme
                | view::split_when(is_escape)
                | view::join_when([](auto&&){ return false; }); // escapes should be {"", "\thell", "\\", "\\\\o"}
   CHECK(distance(escapes) > 0); // `not empty(escapes)` is a compile-time error :-(
}
```

## Snippet B
```cpp
static_assert(ForwardIterator<decltype(current_)>); // current_ is a split_when_view::iterator.

if (current_ == end(base_->outer_)) {
   joined_range_ = subrange<iterator_t<R>>(current_, current_);
   return;
}

auto i = find_if_not(current_, end(base_->outer_), std::ref(base_->pred_), std::ref(base_->proj_));
if (i != end(base_->outer_)) {
   ++i;
   assert(i != current_); // Fires, but it should not.
}
```

**Comment by CaseyCarter** [comment]
I have weird feelings about `join_when::cursor::equal(cursor)` that I commented on the linked commit. If that doesn't help, please produce an [MCVE](https://stackoverflow.com/help/mcve) that reproduces the bug.

**Comment by cjdb** [comment]
> I have weird feelings about `join_when::cursor::equal(cursor)` that I commented on the linked commit. If that doesn't help, please produce an [MCVE](https://stackoverflow.com/help/mcve) that reproduces the bug.

Thanks for spotting _my_ bug. MCVE below.

The assert should not fire: as `i` models a `ForwardIterator`, `++i` should advance the iterator to something other than `begin(escapes)`. Note that I've only been able to reproduce the failure under these conditions:

1. `is_escape` sometimes returns `true` and sometimes returns `false`.
2. I get the next iterator in the sequence and return `false` when `n == last`.

```cpp
#include <range/v3/view/split_when.hpp>
#include <string_view>
#include <cassert>

int main()
{
   using namespace ranges;
   using namespace std::string_view_literals;

   auto is_escape = [](auto first, auto last) {
      if (auto n = next(first); n != last) {
         return std::make_pair(true, first);
      }
      return std::make_pair(false, first);
   };

   auto lexeme = R"(\t)"sv;
   auto escapes = lexeme
                | view::split_when(is_escape);
   static_assert(ForwardRange<decltype(escapes)>);

   if (auto i = begin(escapes); i != end(escapes)) {
      ++i;
      assert(i != begin(escapes));
   }
}
```

**Comment by CaseyCarter** [comment]
TLDR: `split_when_view::cursor`'s constructor was assuming that `is_escape(i, j) == pair(true, i)` implies `is_escape(i, i + 1) == pair(true, i)`. This makes sense for simple pattern matches, but is not the case for more complex delimiter predicates (as you noticed).

**Comment by cjdb** [comment]
Thanks, the patch in 9aff83a fixes this problem!
